urlc5fdb179cc_lagodiuk_evo_neural_network_agents_tgz-pJ8-com_lagodiuk_agent_evolution_MainJ8/src/com/lagodiuk/agent/evolution/NeuralNetworkDrivenAgent.java

```diff
@@ -0,0 +1,255 @@
+	 * brain, while method "interact" runs <br/>
+	 * <br/>
+	 * TODO Maybe consider to use non-blocking technique. But at the moment this
+	 * simplest solution doesn't cause any overheads
+	 */
+	@Override
+	public synchronized void interact(AgentsEnvironment env) {
+		List<Double> nnInputs = this.createNnInputs(env);
+
+		this.activateNeuralNetwork(nnInputs);
+
+		int neuronsCount = this.brain.getNeuronsCount();
+		double deltaAngle = this.brain.getAfterActivationSignal(neuronsCount - 2);
+		double deltaSpeed = this.brain.getAfterActivationSignal(neuronsCount - 1);
+
+		deltaSpeed = this.avoidNaNAndInfinity(deltaSpeed);
+		deltaAngle = this.avoidNaNAndInfinity(deltaAngle);
+
+		double newSpeed = this.normalizeSpeed(this.getSpeed() + deltaSpeed);
+		double newAngle = this.getAngle() + this.normalizeDeltaAngle(deltaAngle);
+
+		this.setAngle(newAngle);
+		this.setSpeed(newSpeed);
+
+		this.move();
+	}
+
+	private double avoidNaNAndInfinity(double x) {
+		if ((Double.isNaN(x)) || Double.isInfinite(x)) {
+			x = 0;
+		}
+		return x;
+	}
+
+	private void activateNeuralNetwork(List<Double> nnInputs) {
+		for (int i = 0; i < nnInputs.size(); i++) {
+			this.brain.putSignalToNeuron(i, nnInputs.get(i));
+		}
+		this.brain.activate();
+	}
+
+	protected List<Double> createNnInputs(AgentsEnvironment environment) {
+		// Find nearest food
+		Food nearestFood = null;
+		double nearestFoodDist = Double.MAX_VALUE;
+
+		for (Food currFood : environment.filter(Food.class)) {
+			// agent can see only ahead
+			if (this.inSight(currFood)) {
+				double currFoodDist = this.distanceTo(currFood);
+				if ((nearestFood == null) || (currFoodDist <= nearestFoodDist)) {
+					nearestFood = currFood;
+					nearestFoodDist = currFoodDist;
+				}
+			}
+		}
+
+		// Find nearest agent
+		Agent nearestAgent = null;
+		double nearestAgentDist = maxAgentsDistance;
+
+		for (Agent currAgent : environment.filter(Agent.class)) {
+			// agent can see only ahead
+			if ((this != currAgent) && (this.inSight(currAgent))) {
+				double currAgentDist = this.distanceTo(currAgent);
+				if (currAgentDist <= nearestAgentDist) {
+					nearestAgent = currAgent;
+					nearestAgentDist = currAgentDist;
+				}
+			}
+		}
+
+		List<Double> nnInputs = new LinkedList<Double>();
+
+		double rx = this.getRx();
+		double ry = this.getRy();
+
+		double x = this.getX();
+		double y = this.getY();
+
+		if (nearestFood != null) {
+			double foodDirectionVectorX = nearestFood.getX() - x;
+			double foodDirectionVectorY = nearestFood.getY() - y;
+
+			// left/right cos
+			double foodDirectionCosTeta =
+					Math.signum(this.pseudoScalarProduct(rx, ry, foodDirectionVectorX, foodDirectionVectorY))
+							* this.cosTeta(rx, ry, foodDirectionVectorX, foodDirectionVectorY);
+
+			nnInputs.add(FOOD);
+			nnInputs.add(nearestFoodDist);
+			nnInputs.add(foodDirectionCosTeta);
+
+		} else {
+			nnInputs.add(EMPTY);
+			nnInputs.add(0.0);
+			nnInputs.add(0.0);
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 144:161, 163:180.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.